[udd/update] Add the `inputFormat` option for supporting the XML/JSON…

kyoshinaga opened the pull request:
… input. And add test files into the src/test/resources/data/{xml, json} directory.

Integration test is done on the [`jigg-pipeline-test`](https://github.com/tomeken-yoshinaga/jigg-pipeline-test.git) repository. The result is shown on the README.

Commit:
[udd/update] Add the `inputFormat` option for supporting the XML/JSON input. And add test files into the src/test/resources/data/{xml, json} directory.

hiroshinoji commented on src/main/scala/jigg/pipeline/Pipeline.scala:
```diff
@@ -148,15 +150,13 @@ Currently the annotators listed below are installed. See the detail of each anno
-            println("***** WARNING *****")
-            println(e.getMessage)
-            println("*******************")
+            System.out.println(s"[warn] %s".format(e.getMessage))
```
Sorry for comments on previous commit. Could you remove duplications in code by higher-order function? E.g.,

``` scala
    def check(procError: (RequirementError, RequirementSet)=>RequirementSet) =
      annotatorList.foldLeft(RequirementSet()) { (satisfiedSoFar, annotator) =>
        try annotator.checkRequirements(satisfiedSoFar)
        catch { case e: RequirementError => procError(e, satisfiedSoFar) }
      }

    checkRequirement match {
      case "true" => check { (e, satisfiedSoFar) =>
        argumentError("annotators", e.getMessage)
      }
      case "warn" => check { (e, satisfiedSoFar) =>
        System.out.println(s"[warn] %s".format(e.getMessage))
        satisfiedSoFar
      }
...
```

This may also highlight the difference between two options.

kyoshinaga replied:
Thanks a lot for this suggestion.
Ok, I'm going to modify this part.

hiroshinoji commented on src/main/scala/jigg/pipeline/Pipeline.scala:
```diff
@@ -148,15 +150,13 @@ Currently the annotators listed below are installed. See the detail of each anno
       }
       case "false" => {
-        println("Skip `checkRequirements` for all annotators.")
+        System.out.println("[warn] SKIP `checkRequirements` for all selected annotators.")
```
Outputting messages in standard output may cause errors since some users may connect the output of jigg in pipe.

kyoshinaga replied:
Sorry, this is my fault.
I'm going to modify it to output in standard error.

hiroshinoji commented on src/main/scala/jigg/util/JSONUtil.scala:
```diff
@@ -90,10 +92,18 @@ object JSONUtil {
+        // To avoid this issues, all such strings is unescaped before throw into
+        // the XMLUtil.addAttributes method.
+        .map { case (k, v) =>
+        (k, unescapeMap.foldLeft(v.toString){(text, pair) => text.replaceAll(pair._1, pair._2)})}
```
This `foldLeft` and the next similar line (106) could be extracted by a function?

Commit:
[update] Modify following codes.

- Pipeline.createAnnotatorList:
  - Remove duplications by higher-order `check` function.
  - Output message in standard error.
- XMLUtil:
  - Rename `getChildNode` to `getNonEmptyChild`.
- JSONUtil:
  - Extract `foldLeft` line by higher-order `replaceAllEscape` function.

hiroshinoji commented on src/main/scala/jigg/util/XMLUtil.scala:
```diff
@@ -119,13 +119,13 @@ object XMLUtil {
-  def getChildNode(nodes: NodeSeq): NodeSeq = nodes match {
-    case x: Elem => x.child filter (_.label != "#PCDATA")
-    case x => (x.map(getChildNode(_))).flatten
+  def getNonEmptyChild(nodes: NodeSeq): NodeSeq = nodes match {
```
This and `hasChild` method may be very inefficient when the size of XML tree is huge. Is it possible to avoid traversing all nodes in the tree?

hiroshinoji commented on src/main/scala/jigg/util/XMLUtil.scala:
```diff
@@ -134,4 +134,29 @@ object XMLUtil {
+        case t => <xml>{t}</xml>
+      }
+      val children = n match {
+        case n if getNonEmptyChild(n).filterNot(x => hasChild(x)).length == 0 =>
```
This statement may fail when only a part of child elements have non-empty child, for example

``` xml
<sentence>
  <tokens>...</tokens>
  <NEs/>
</sentence>
```

Commit:
[update] Fix bug.

Commit:
[update] Delete debugging line.

kyoshinaga commented:
Sorry, a debugging output was not deleted in `XMLUtilSpec`.

hiroshinoji commented:
LGTM. Thank you!
